From df41f91c6de52d1715fc3dd5d1a3e73dcc71722a Mon Sep 17 00:00:00 2001
From: Sammy He <r62914@freescale.com>
Date: Mon, 20 Sep 2010 22:32:21 +0800
Subject: [PATCH 081/635] ENGR00131746 Add more protection in semaphore wait
 to avoid lock

Add more protection in semaphore wait to avoid lock, this is done
in CaptureFrameThread.

Signed-off-by: Sammy He <r62914@freescale.com>
---
 libcamera/CameraHal.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/libcamera/CameraHal.cpp b/libcamera/CameraHal.cpp
index af1a841..11e712e 100755
--- a/libcamera/CameraHal.cpp
+++ b/libcamera/CameraHal.cpp
@@ -764,7 +764,10 @@ int CameraHal::previewCaptureFrameThread()
         do {
             clock_gettime(CLOCK_REALTIME, &ts);
             ts.tv_nsec +=100000; // 100ms
-        } while ((sem_timedwait(&avaible_dequeue_frame, &ts) != 0) && !error_status);
+        } while ((sem_timedwait(&avaible_dequeue_frame, &ts) != 0) && mPreviewRunning && !error_status);
+
+        if ((mPreviewRunning == 0) || error_status)
+            return -1;
 
 	memset(&cfilledbuffer, 0, sizeof (cfilledbuffer));
 	cfilledbuffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
-- 
1.8.0

